### PR TITLE
Add support for sub-indexes

### LINKED ISSUE #31: Add  support for sub-indexes
Packages that have subpackages, like `net/` in `net/metrics` currently 404.

Instead, we should render the same index that is shown at `/`, but scoped to that package.

For example, given the packages:

* `go.uber.org/net/metrics`
* `go.uber.org/net/auth`

Navigating to https://go.uber.org/net would show an index containing the above 2 packages.

### PR DESCRIPTION
**Pull request by abhinav**
<details>
<summary>Background</summary>

Sally renders two kinds of pages:

- packages: These are for packages defined in sally.yaml
  and any route under the package path.
- indexes: These list available packages.

The latter--indexes was previously only supported at '/', the root page.

This leads to a slight UX issue:
if you have a package with a / in its name (e.g. net/metrics):

- example.com/net/metrics gives you the package page
- example.com/ lists net/metrics
- However, example.com/net fails with a 404

</details>

This adds support for index pages on all parents of package pages.
Therefore, if example.com/net/metrics exists,
example.com/net will list all packages defined under that path.

Resolves #31

### CODE REVIEW

**Comment by abhinav** [comment]
The implementation has been simplified significantly.
I've eliminated the custom routing backend in favor of the same old ServeMux-backed router.
The performance difference from the custom router was non-existent and the cost in terms of complexity was a lot higher.

File: handler.go [hunk 1]
```diff
@@ -17,112 +19,169 @@ var (
...
+//	GET /<dir>
+//		Page listing packages under the given directory,
+//		assuming that there's no package with the given name.
+//	GET /<name>/<subpkg>
+//		Package page for the given subpackage.
 func CreateHandler(config *Config) http.Handler {
 	mux := http.NewServeMux()
-
-	pkgs := make([]packageInfo, 0, len(config.Packages))
+	pkgs := make([]*sallyPackage, 0, len(config.Packages))
```
Comment (mway) [inline handler.go:37]: nit: `var` group

File: handler.go [hunk 2]
```diff
@@ -17,112 +19,169 @@ var (
...
+		modulePath := path.Join(baseURL, name)
+		docURL := "https://" + path.Join(config.Godoc.Host, modulePath)
+
+		pkg := &sallyPackage{
+			Name:       name,
+			Desc:       pkg.Desc,
+			ModulePath: modulePath,
+			DocURL:     docURL,
+			GitURL:     pkg.Repo,
+		}
```
Comment (mway) [inline handler.go:53]: nit: grouping

```suggestion

		var (
			modulePath = path.Join(baseURL, name)
			docURL     = "https://" + path.Join(config.Godoc.Host, modulePath)
			pkg        = &sallyPackage{
				Name:       name,
				Desc:       pkg.Desc,
				ModulePath: modulePath,
				DocURL:     docURL,
				GitURL:     pkg.Repo,
			}
		)
```

File: handler.go [hunk 3]
```diff
@@ -17,112 +19,169 @@ var (
...
+	err := packageTemplate.Execute(w,
+		struct {
+			ModulePath string
+			GitURL     string
+			DocURL     string
+		}{
+			ModulePath: h.Pkg.ModulePath,
+			GitURL:     h.Pkg.GitURL,
+			DocURL:     h.Pkg.DocURL + relPath,
+		})
```
Comment (mway) [inline handler.go:179]: nit: formatting
```suggestion
	err := packageTemplate.Execute(w, struct {
		ModulePath string
		GitURL     string
		DocURL     string
	}{
		ModulePath: h.Pkg.ModulePath,
		GitURL:     h.Pkg.GitURL,
		DocURL:     h.Pkg.DocURL + relPath,
	})
```

**Review by mway — APPROVED** [review]
LGTM

**Review by sywhang — APPROVED** [review]

File: handler.go (hunk 1 above)
Comment (abhinav) [reply handler.go:37]: I agree with var grouping when `var` form is used, but I think when we're initializing with non-zero values, var form is rarely used.

Per the style guide, we use var form for zero value initialization and := form for non-zero-value:
https://github.com/uber-go/guide/blob/master/style.md#use-var-for-zero-value-structs
https://github.com/uber-go/guide/blob/master/style.md#initializing-struct-references

(That links only talk about structs, but the guidance has been consistent on this.)

The readability in a var group suffers as soon as you add a type for any actual var-form-zero-values.

```
var (
  mux           = http.NewServeMux()
  pkgs          = make(...)
  foo io.Reader
)
```

File: handler.go (hunk 2 above)
Comment (abhinav) [reply handler.go:53]: Same comment as above. The var form is just less readable here. `:=` for non-zero value initialization has always been the way.

**Review by abhinav — COMMENTED** [review]
Thanks for the reviews, @mway @sywhang.
I've applied some suggestions, and left comments on a couple.
Let me know what you think.

File: handler.go (hunk 1 above)
Comment (mway) [reply handler.go:37]: It's fine as-is. :) Not trying to bikeshed.

My point was more that I don't personally see the value in segmenting the initializations here so I would probably group them all. I disagree about var group readability suffering when mixing initialization and zero-value declarations (and I think that ultimately leads to a lot more ad hoc initialization/declaration patterns).

**Comment by abhinav** [comment]
Per chat with @mway, will leave `:=` initialization in place for cases where var form isn't used because a type isn't specified.

**Pull request merged**